src/neuroglancer/sliceview/png/index.ts

```diff
@@ -0,0 +1,213 @@
+/**
+ * @license
+ * Copyright 2022 William Silvermsith
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import pngWasmDataUrl from './libpng.wasm';
+
+const libraryEnv = {
+  emscripten_notify_memory_growth: function () {},
+  proc_exit: (code: number) => {
+    throw `proc exit: ${code}`;
+  },
+};
+
+const pngModulePromise = (async () => {
+  const response = await fetch(pngWasmDataUrl);
+  const wasmCode = await response.arrayBuffer();
+  const m = await WebAssembly.instantiate(wasmCode, {
+    env: libraryEnv,
+    wasi_snapshot_preview1: libraryEnv,
+  });
+  (m.instance.exports._initialize as Function)();
+  return m;
+})();
+
+const enum PngColorSpace {
+  GRAYSCALE = 0,
+  RGB = 2,
+  PALETTE = 3,
+  GRAYSCALE_ALPHA = 4,
+  RGBA = 6
+}
+
+// header constants
+const magicSpec = [ 137, 80, 78, 71, 13, 10, 26, 10 ];
+const validHeaderCode = [ 'I', 'H', 'D', 'R' ];
+
+// not a full implementation of read header, just the parts we need
```

Comment:
I know you put in some export to implement this header parsing in javascript, but is there any advantage over just relying on spng to do that?

Reply:
The main advantage I can think of is that the error message is easier to write in an informative way. I can revert it back if you'd like.